modules/stereo/src/matching.cpp

```diff
@@ -0,0 +1,338 @@
+//warranties of merchantability and fitness for a particular purpose are disclaimed.
+//In no event shall copyright holders or contributors be liable for any direct,
+//indirect, incidental, special, exemplary, or consequential damages
+//(including, but not limited to, procurement of substitute goods or services;
+//loss of use, data, or profits; or business interruption) however caused
+//and on any theory of liability, whether in contract, strict liability,
+//or tort (including negligence or otherwise) arising in any way out of
+//the use of this software, even if advised of the possibility of such damage.
+#include "precomp.hpp"
+#include "matching.hpp"
+
+namespace cv
+{
+    namespace stereo
+    {
+        //null constructor
+        Matching::Matching(void)
+        {
+        }
+        Matching::~Matching(void)
+        {
+        }
+        //constructor for the matching class
+        //maxDisp - represents the maximum disparity
+        Matching::Matching(int maxDisp, int scalling, int confidence)
+        {
+            CV_Assert(maxDisp > 10);
+            CV_Assert(scalling != 0);
+            CV_Assert(confidence >= 1);
+            this->scallingFactor = scalling;
+            //set the maximum disparity
+            this->maxDisparity = maxDisp;
+            //set the value for the confidence
+            this->confidenceCheck = confidence;
+            //generate the hamming lut in case SSE is not available
+            hammingLut();
+        }
+        //!method for setting the maximum disparity
+        void Matching::setMaxDisparity(int val)
+        {
+            CV_Assert(val > 10);
+            this->maxDisparity = val;
+        }
+        //!method for getting the disparity
+        int Matching::getMaxDisparity()
+        {
+            return this->maxDisparity;
+        }
+        void Matching::setScallingFactor(int val)
+        {
+            CV_Assert(val > 0);
+            scallingFactor = val;
+        }
+        //!method for getting the scalling factor
+        int Matching::getScallingFactor()
+        {
+            return scallingFactor;
+        }
+        //! Hamming distance computation method
+        //! leftImage and rightImage are the two transformed images
+        //! the cost is the resulted cost volume and kernel Size is the size of the matching window
+        void Matching::hammingDistanceBlockMatching(const Mat &leftImage, const Mat &rightImage, Mat &cost, const int kernelSize)
+        {
+            CV_Assert(leftImage.cols == rightImage.cols);
+            CV_Assert(leftImage.rows == rightImage.rows);
+            CV_Assert(kernelSize % 2 != 0);
+            CV_Assert(cost.rows == leftImage.rows);
+            CV_Assert(cost.cols / (maxDisparity + 1) == leftImage.cols);
+            // cost.setTo(0);
+            int *c = (int *)cost.data;
+            memset(c, 0, sizeof(c[0]) * leftImage.cols * leftImage.rows * (maxDisparity + 1));
```

Comment:
Sorry, Paul, I forgot, what's wrong with setTo?

Also, you're using .cols instead of stride here as well. It's probably, not exactly an error right here, but anyway, later on it can shoot you in the foot. That's why I'm saying that messing directly with pointers is error-prone, you have to be very careful. Once again, I would suggest to use "politically correct" OpenCV ways of doing things to avoid such errors (like, setTo in this case). 99% of the time there are no good reasons to not do this.